game-core/src/main/java/games/strategy/triplea/ai/pro/util/ProTerritoryValueUtils.java

```diff
@@ -429,18 +431,41 @@ private static double findWaterValue(
     return capitalOrFactoryValue / 100 + nearbyLandValue / 10;
   }
 
+  // 1700 oldT=4906 newT=227 newT2=166 newT3=152
+  @SuppressWarnings("ReferenceEquality")
   private static Set<Territory> findNearbyEnemyCapitalsAndFactories(
-      final Territory t, final Map<Territory, Double> enemyCapitalsAndFactoriesMap) {
-
-    Set<Territory> nearbyEnemyCapitalsAndFactories = new HashSet<>();
-    for (int i = MIN_FACTORY_CHECK_DISTANCE; i <= MAX_FACTORY_CHECK_DISTANCE; i++) {
-      nearbyEnemyCapitalsAndFactories = t.getData().getMap().getNeighbors(t, i);
-      nearbyEnemyCapitalsAndFactories.retainAll(enemyCapitalsAndFactoriesMap.keySet());
-      if (!nearbyEnemyCapitalsAndFactories.isEmpty()) {
-        break;
+      final Territory t, final Set<Territory> enemyCapitalsAndFactories) {
+    final GameMap map = t.getData().getMap();
+
+    int distance = 0;
+    Territory nextDistanceAfter = t;
+
+    final var found = new HashSet<Territory>();
```

Comment:
IDK, I'd tend to agree with code climate on the length and complexity warning. The trick is not just to break up a method but to create the right abstractions. EG:

```
while (!queue.isEmpty()) {
   found.addAll(findAtDistance(distance, .... ));
   distance ++;
   if (distance > MAX_FACTORY_CHECK_DISTANCE
        || (distance >= MIN_FACTORY_CHECK_DISTANCE && !found.isEmpty())) {
       break;
    }
    nextDistanceAfter = queue.getLast();
}
return found;
```

I don't know if that quite does it, it's an art to find something that works right and does not just create indirection.

Reply:
I don't think that helps, findAtDistance() would require too many params which codeclimate will complain about.

The only way I see of "working around it" is to make a helper class and move all these variables to be members. Then, helper methods don't need to take these params because they'd be explicitly taking them as part of being on the object, so codeclimate would not complain.

I'm not sure this is something we should be optimizing for. The result would be instead of a 30-line method we'd have a 50-75 line class. Is that really a net win?

Reply:
Put another way, would we be happy if every 30+ line method instead becomes its own class with a constructor and 2-5 other methods? Do we think that would be a better state?

Reply:
I'd say the complexity is ok here given this is a performance sensitive method. If anything I'd rather focus on adding some comments describing how it works as no matter how you chop it up it is going to be a bit complex.

Reply:
I don't think that is necessarily the trade-off or decision. There are other ways..

Reply:
> I don't think that helps, findAtDistance() would require too many params which codeclimate will complain about.

[Introduce Parameter Object](https://refactoring.com/catalog/introduceParameterObject.html) refactoring would solve that.

The big point is it's not necessarily about creating indirection, but abstractions and information hiding.

Reply:
To expound slightly,  a data object with `@Builder` makes for a natural parameter object.

Reply:
One simplification I made is removing MAX_FACTORY_CHECK_DISTANCE.

This is technically a change from the previous implementation, but in practice I think the only reason the previous implementation had it is because such a condition was needed to make sure it terminated (the old algo just expanded distance and kept trying to find things).

It's not needed in this implementation and just adds extra complexity. In practice, 30 away is very far. A small map like revised seems to have every territory less than 9 territories away from any other. A larger map like Big World still is well within the 30 limit.